README.pod
=pod

=head1 NAME

Time::FFI - libffi interface to POSIX date and time functions

=head1 SYNOPSIS

  use Time::FFI qw(localtime mktime strptime strftime);

  my $tm = strptime '1995-01-02 13:15:39', '%Y-%m-%d %H:%M:%S';
  my $epoch = mktime $tm;
  print "$epoch: ", strftime('%I:%M:%S %p on %B %e, %Y', $tm);

  my $tm = localtime time;
  my $datetime = $tm->to_object_as_local('DateTime');

  my $tm = gmtime time;
  my $moment = $tm->to_object_as_utc('Time::Moment');

  use Time::FFI::tm;
  my $tm = Time::FFI::tm->from_object(DateTime->now);
  my $epoch = $tm->epoch_as_local;
  my $piece = $tm->to_object_as_local('Time::Piece');

=head1 DESCRIPTION

B<Time::FFI> provides a L<libffi|FFI::Platypus> interface to POSIX date and
time functions found in F<time.h>.

The L</gmtime> and L</localtime> functions behave very differently from the
core functions of the same name, as well as those exported by L<Time::Piece>,
so you may wish to call them as e.g. C<Time::FFI::gmtime> rather than importing
them.

All functions will throw an exception in the event of an error. For functions
other than L</strftime> and L</strptime>, this exception will contain the
syscall error message, and L<perlvar/$!> will also have been set by the
syscall, so you could check it after trapping the exception for finer exception
handling.

=head1 FUNCTIONS

All functions are exported individually, or with the C<:all> export tag.

=head2 asctime

  my $str = asctime $tm;

Returns a string in the format C<Wed Jun 30 21:49:08 1993\n> representing the
passed L<Time::FFI::tm> record. The thread-safe L<asctime_r(3)> function is
used if available.

=head2 ctime

  my $str = ctime $epoch;
  my $str = ctime;

Returns a string in the format C<Wed Jun 30 21:49:08 1993\n> representing the
passed epoch timestamp (defaulting to the current time) in the local time zone.
This is equivalent to L<POSIX/ctime> but uses the thread-safe L<ctime_r(3)>
function if available.

=head2 gmtime

  my $tm = gmtime $epoch;
  my $tm = gmtime;

Returns a L<Time::FFI::tm> record representing the passed epoch timestamp
(defaulting to the current time) in UTC. The thread-safe L<gmtime_r(3)>
function is used if available.

=head2 localtime

  my $tm = localtime $epoch;
  my $tm = localtime;

Returns a L<Time::FFI::tm> record representing the passed epoch timestamp
(defaulting to the current time) in the local time zone. The thread-safe
L<localtime_r(3)> function is used if available.

=head2 mktime

  my $epoch = mktime $tm;

Returns the epoch timestamp representing the passed L<Time::FFI::tm> record
interpreted in the local time zone. The time is interpreted from the C<sec>,
C<min>, C<hour>, C<mday>, C<mon>, C<year>, and C<isdst> members of the record,
ignoring the rest. DST status will be automatically determined if C<isdst> is a
negative value. The record will also be updated to normalize any out-of-range
values and populate the C<isdst>, C<wday>, and C<yday> values, as well as
C<gmtoff> and C<zone> if supported.

=head2 strftime

  my $str = strftime $format, $tm;

Returns a string formatted according to the passed format string, representing
the passed L<Time::FFI::tm> record. Consult your system's L<strftime(3)> manual
for available format descriptors.

=head2 strptime

  my $tm = strptime $str, $format;
     $tm = strptime $str, $format, $tm;
  my $tm = strptime $str, $format, undef, \my $remaining;
     $tm = strptime $str, $format, $tm, \my $remaining;

Returns a L<Time::FFI::tm> record representing the passed string, parsed
according to the passed format. Consult your system's L<strptime(3)> manual for
available format descriptors. The C<isdst> value will be set to -1; all other
unspecified values will default to 0. Note that the default C<mday> value of 0
is outside of the standard range [1,31] and may cause an error or be
interpreted as the last day of the previous month.

A L<Time::FFI::tm> record may be passed as the third argument, in which case it
will be modified in place to (on most systems) update only the date/time
elements which were parsed from the string. Additionally, an optional scalar
reference may be passed as the fourth argument, in which case it will be set to
the remaining unprocessed characters of the input string if any.

This function is usually not available on Windows.

=head2 timegm

  my $epoch = timegm $tm;

I<Since version 1.002>

Like L</mktime>, but interprets the passed L<Time::FFI::tm> record as UTC. This
function is not always available.

=head2 timelocal

  my $epoch = timelocal $tm;

I<Since version 1.002>

The same as L</mktime>, but not always available.

=head1 BUGS

Report any issues on the public bugtracker.

=head1 AUTHOR

Dan Book <dbook@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is Copyright (c) 2019 by Dan Book.

This is free software, licensed under:

  The Artistic License 2.0 (GPL Compatible)

=head1 SEE ALSO

L<Time::Piece>, L<Time::Moment>, L<DateTime>, L<POSIX>, L<POSIX::strptime>

=cut


Changes
{{$NEXT}}

2.002     2021-07-09 02:56:26 EDT
  - Added to_object_as_local, to_object_as_utc, epoch_as_local, epoch_as_utc,
    normalized_as_local, and normalized_as_utc methods to Time::FFI::tm.
  - Deprecated to_object, epoch, and normalized methods for Time::FFI::tm.

2.001     2019-11-27 13:09:36 EST
  - Convert to use FFI::Platypus type parser version 1, no functional changes.

2.000     2019-10-18 18:26:15 EDT
  - Reworked the from_object constructor for Time::FFI::tm to always create a
    tm record based on the localtime attributes of the passed object. For the
    previous absolute time behavior, retrieve an epoch timestamp from the
    original object and pass it to localtime or gmtime to create a
    Time::FFI::tm record.
  - Renamed Time::FFI::tm attributes to remove superfluous tm_. Aliases with
    the tm_ prefix remain.
  - Allow Time::FFI::tm conversion to or from Time::tm and Time::FFI::tm.
  - Preserve class from invocant in Time::FFI::tm normalized method.

1.004     2019-10-10 18:27:29 EDT
  - Default tm_mday to 0 in strptime for consistency with other fields on
    Solaris.

1.003     2019-10-10 17:33:23 EDT
  - Clarify that mktime may adjust any elements of the passed tm record.
  - Renamed with_extra method for Time::FFI::tm to normalized and clarify that
    it also normalizes out-of-range values when interpreting as local time.
  - Allow epoch and to_object methods for Time::FFI::tm to normalize
    out-of-range values when interpreting as local time.

1.002     2019-10-10 16:17:27 EDT
  - Added timelocal and timegm functions where available.

1.001     2019-10-10 13:58:11 EDT
  - Added from_object constructor for Time::FFI::tm.

1.000     2019-10-10 12:19:57 EDT
  - Fixed bugs in Time::FFI::tm->to_object where the tm object may be mutated
    and the results may be incorrect when interpreted in the local time zone.
  - Documented that the tm_mday and tm_isdst members of a new Time::FFI::tm
    record should be set explicitly.
  - Documented that mktime adjusts the tm_isdst, tm_wday, tm_yday, and possibly
    tm_gmtoff and tm_zone members of the passed tm record.
  - Initialize tm_mday to 1 and tm_isdst to -1 in the tm record returned by
    strptime, rather than the defaults of 0.
  - Added epoch method to Time::FFI::tm to return the associated Unix epoch
    timestamp.
  - Added with_extra method to Time::FFI::tm to return a new tm record with
    extra values populated.

0.002     2019-07-05 12:42:42 EDT
  - Skip strptime tests when the function is not provided by the system

0.001     2019-07-03 22:33:35 EDT
  - First release


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.012.
use strict;
use warnings;

use 5.008001;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "libffi interface to POSIX date and time functions",
  "AUTHOR" => "Dan Book <dbook\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Time-FFI",
  "LICENSE" => "artistic_2",
  "MIN_PERL_VERSION" => "5.008001",
  "NAME" => "Time::FFI",
  "PREREQ_PM" => {
    "Carp" => 0,
    "Exporter" => "5.57",
    "FFI::Platypus" => 0,
    "Module::Runtime" => 0,
    "Time::Local" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "Test2::V0" => 0,
    "Test::More" => 0,
    "Test::Needs" => 0
  },
  "VERSION" => "1.005",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Carp" => 0,
  "Exporter" => "5.57",
  "ExtUtils::MakeMaker" => 0,
  "FFI::Platypus" => 0,
  "File::Spec" => 0,
  "Module::Runtime" => 0,
  "Test2::V0" => 0,
  "Test::More" => 0,
  "Test::Needs" => 0,
  "Time::Local" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);


lib/Time/FFI/tm.pm
package Time::FFI::tm;

use strict;
use warnings;
use Carp ();
use FFI::Platypus::Record ();
use Module::Runtime ();
use Time::Local ();

our $VERSION = '2.003';

my @tm_members = qw(sec min hour mday mon year wday yday isdst);

FFI::Platypus::Record::record_layout_1(
  (map { (int => $_) } @tm_members),
  long   => 'gmtoff',
  string => 'zone',
);

{
  no strict 'refs';
  *{"tm_$_"} = \&$_ for @tm_members, 'gmtoff', 'zone';
}

sub from_list {
  my ($class, @args) = @_;
  my %attr = map { ($tm_members[$_] => $args[$_]) } 0..$#tm_members;
  return $class->new(\%attr);
}

sub from_object {
  my ($class, $obj) = @_;
  if ($obj->isa('Time::Piece')) {
    return $class->new(
      year  => $obj->year - 1900,
      mon   => $obj->mon - 1,
      mday  => $obj->mday,
      hour  => $obj->hour,
      min   => $obj->min,
      sec   => $obj->sec,
      isdst => -1,
    );
  } elsif ($obj->isa('Time::Moment')) {
    return $class->new(
      year  => $obj->year - 1900,
      mon   => $obj->month - 1,
      mday  => $obj->day_of_month,
      hour  => $obj->hour,
      min   => $obj->minute,
      sec   => $obj->second,
      isdst => -1,
    );
  } elsif ($obj->isa('DateTime')) {
    return $class->new(
      year  => $obj->year - 1900,
      mon   => $obj->month - 1,
      mday  => $obj->day,
      hour  => $obj->hour,
      min   => $obj->minute,
      sec   => $obj->second,
      isdst => -1,
    );
  } elsif ($obj->isa('Time::FFI::tm') or $obj->isa('Time::tm')) {
    my %attr = map { ($_ => $obj->$_) } qw(sec min hour mday mon year wday yday isdst);
    return $class->new(\%attr);
  } else {
    my $class = ref $obj;
    Carp::croak "Cannot convert from unrecognized object class $class";
  }
}

sub to_list {
  my ($self) = @_;
  return map { $self->$_ } @tm_members;
}

sub to_object {
  Carp::carp '->to_object is deprecated; use ->to_object_as_local or ->to_object_as_utc';
  return _to_object(@_);
}

sub to_object_as_local {
  my ($self, $class) = @_;
  return _to_object($self, $class, 1);
}

sub to_object_as_utc {
  my ($self, $class) = @_;
  return _to_object($self, $class, 0);
}

sub _to_object {
  my ($self, $class, $islocal) = @_;
  Module::Runtime::require_module $class;
  if ($class->isa('Time::Piece')) {
    my ($epoch) = $islocal ? _mktime($self) : _timegm($self);
    return $islocal ? scalar $class->localtime($epoch) : scalar $class->gmtime($epoch);
  } elsif ($class->isa('Time::Moment')) {
    my $tm = $self;
    my $epoch;
    ($epoch, $tm) = _mktime($self) if $islocal;
    my $moment = $class->new(
      year   => $tm->year + 1900,
      month  => $tm->mon + 1,
      day    => $tm->mday,
      hour   => $tm->hour,
      minute => $tm->min,
      second => $tm->sec,
    );
    return $islocal ? $moment->with_offset_same_local(($moment->epoch - $epoch) / 60) : $moment;
  } elsif ($class->isa('DateTime')) {
    my $tm = $self;
    (undef, $tm) = _mktime($self) if $islocal;
    return $class->new(
      year   => $tm->year + 1900,
      month  => $tm->mon + 1,
      day    => $tm->mday,
      hour   => $tm->hour,
      minute => $tm->min,
      second => $tm->sec,
      time_zone => $islocal ? 'local' : 'UTC',
    );
  } elsif ($class->isa('Time::FFI::tm') or $class->isa('Time::tm')) {
    my %attr = map { ($_ => $self->$_) } qw(sec min hour mday mon year wday yday isdst);
    return $class->new(%attr);
  } else {
    Carp::croak "Cannot convert to unrecognized object class $class";
  }
}

sub epoch {
  my ($self, $islocal) = @_;
  Carp::carp '->epoch is deprecated; use ->epoch_as_local or ->epoch_as_utc';
  my ($epoch) = $islocal ? _mktime($self) : _timegm($self);
  return $epoch;
}

sub epoch_as_local {
  my ($self) = @_;
  my ($epoch) = _mktime($self);
  return $epoch;
}

sub epoch_as_utc {
  my ($self) = @_;
  my ($epoch) = _timegm($self);
  return $epoch;
}

sub normalized {
  my ($self, $islocal) = @_;
  Carp::carp '->normalized is deprecated; use ->normalized_as_local or ->normalized_as_utc';
  return $islocal ? $self->normalized_as_local : $self->normalized_as_utc;
}
*with_extra = \&normalized;

sub normalized_as_local {
  my ($self) = @_;
  my (undef, $new) = _mktime($self);
  return $new;
}

sub normalized_as_utc {
  my ($self) = @_;
  my ($epoch) = _timegm($self);
  require Time::FFI;
  my $new = Time::FFI::gmtime($epoch);
  bless $new, ref $self;
  return $new;
}

sub _mktime {
  my ($self) = @_;
  require Time::FFI;
  my %attr = map { ($_ => $self->$_) } qw(sec min hour mday mon year);
  $attr{isdst} = -1;
  my $new = (ref $self)->new(\%attr);
  return (Time::FFI::mktime($new), $new);
}

sub _timegm {
  my ($self) = @_;
  my $year = $self->year;
  $year += 1900 if $year >= 0; # avoid timegm year heuristic
  my @vals = ((map { $self->$_ } qw(sec min hour mday mon)), $year);
  return scalar Time::Local::timegm(@vals);
}

1;

=head1 NAME

Time::FFI::tm - POSIX tm record structure

=head1 SYNOPSIS

  use Time::FFI::tm;

  my $tm = Time::FFI::tm->new(
    year  => 95, # years since 1900
    mon   => 0,  # 0 == January
    mday  => 1,
    hour  => 13,
    min   => 25,
    sec   => 59,
    isdst => -1, # allow DST status to be determined by the system
  );
  $tm->mday($tm->mday + 1); # add a day

  my $in_local = $tm->normalized_as_local;
  say $in_local->isdst; # now knows if DST is active

  my $tm = Time::FFI::tm->from_list(CORE::localtime(time));

  my $epoch = POSIX::mktime($tm->to_list);
  my $epoch = $tm->epoch_as_local;

  my $tm = Time::FFI::tm->from_object(Time::Moment->now);
  my $datetime = $tm->to_object_as_local('DateTime');

=head1 DESCRIPTION

This L<FFI::Platypus::Record> class represents the C<tm> struct defined by
F<time.h> and used by functions such as L<mktime(3)> and L<strptime(3)>. This
is used by L<Time::FFI> to provide access to such structures.

The structure does not store an explicit time zone, so you must specify whether
to interpret it as local or UTC time whenever rendering it to an actual
datetime.

=head1 ATTRIBUTES

The integer components of the C<tm> struct are stored as settable attributes
that default to 0.

Note that 0 is out of the standard range for the C<mday> value (often
indicating the last day of the previous month), and C<isdst> should be set to a
negative value if unknown, so these values should always be specified
explicitly.

Each attribute also has a corresponding alias starting with C<tm_> to match the
standard C<tm> struct member names.

=head2 sec

Seconds [0,60].

=head2 min

Minutes [0,59].

=head2 hour

Hour [0,23].

=head2 mday

Day of month [1,31].

=head2 mon

Month of year [0,11].

=head2 year

Years since 1900.

=head2 wday

Day of week [0,6] (Sunday =0).

=head2 yday

Day of year [0,365].

=head2 isdst

Daylight Savings flag. (0: off, positive: on, negative: unknown)

=head2 gmtoff

Seconds east of UTC. (May not be available on all systems)

=head2 zone

Timezone abbreviation. (Read only string, may not be available on all systems)

=head1 METHODS

=head2 new

  my $tm = Time::FFI::tm->new;
  my $tm = Time::FFI::tm->new(year => $year, ...);
  my $tm = Time::FFI::tm->new({year => $year, ...});

Construct a new B<Time::FFI::tm> object representing a C<tm> struct.

=head2 from_list

  my $tm = Time::FFI::tm->from_list($sec, $min, $hour, $mday, $mon, $year, $wday, $yday, $isdst);

Construct a new B<Time::FFI::tm> object from the passed list of time
attributes, in the same order returned by L<perlfunc/localtime>. Missing or
undefined values will be interpreted as the default of 0, but see
L</ATTRIBUTES>.

=head2 from_object

  my $tm = Time::FFI::tm->from_object($obj);

I<Current API since version 2.000.>

Construct a new B<Time::FFI::tm> object from the passed datetime object's local
datetime components. Currently L<Time::Piece>, L<Time::Moment>, L<DateTime>,
L<Time::tm>, and L<Time::FFI::tm> objects (and subclasses) are recognized. The
original time zone and any fractional seconds will not be represented in the
resulting structure.

=head2 to_list

  my ($sec, $min, $hour, $mday, $mon, $year, $wday, $yday, $isdst) = $tm->to_list;

Return the list of time attributes in the structure, in the same order returned
by L<perlfunc/localtime>.

=head2 to_object_as_local

=head2 to_object_as_utc

  my $piece    = $tm->to_object_as_local('Time::Piece');
  my $moment   = $tm->to_object_as_utc('Time::Moment');

I<Since version 2.002.>

Return an object of the specified class. Currently L<Time::Piece>,
L<Time::Moment>, and L<DateTime> (or subclasses) are recognized. Depending on
the method called, the time attributes are interpreted in the local time zone
or in UTC.

When interpreted as a local time, values outside the standard ranges are
accepted; this is not currently supported for UTC times.

You may also specify L<Time::tm> or L<Time::FFI::tm> (or subclasses), in which
case C<to_object_as_local> and C<to_object_as_utc> produce the same result with
the time attributes copied as-is.

=head2 epoch_as_local

=head2 epoch_as_utc

  my $epoch = $tm->epoch_as_local;
  my $epoch = $tm->epoch_as_utc

I<Since version 2.002.>

Translate the time structure into a Unix epoch timestamp (seconds since
1970-01-01 UTC). Depending on the method called, the time attributes are
interpreted in the local time zone or in UTC.

When interpreted as a local time, values outside the standard ranges are
accepted; this is not currently supported for UTC times.

=head2 normalized_as_local

=head2 normalized_as_utc

  my $new = $tm->normalized_as_local;
  my $new = $tm->normalized_as_utc;

I<Since version 2.002.>

Return a new B<Time::FFI::tm> object representing the same time, but with
C<wday>, C<yday>, C<isdst>, and (if supported) C<gmtoff> and C<zone> set
appropriately. Depending on the method called, the time attributes are
interpreted in the local time zone or in UTC.

When interpreted as a local time, values outside the standard ranges will also
be normalized; this is not currently supported for UTC times.

=head1 BUGS

Report any issues on the public bugtracker.

=head1 AUTHOR

Dan Book <dbook@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is Copyright (c) 2019 by Dan Book.

This is free software, licensed under:

  The Artistic License 2.0 (GPL Compatible)

=head1 SEE ALSO

L<Time::FFI>, L<Time::tm>

=for Pod::Coverage with_extra to_object epoch normalized tm_sec tm_min tm_hour tm_mday tm_mon tm_year tm_wday tm_yday tm_isdst tm_gmtoff tm_zone


lib/Time/FFI.pm
package Time::FFI;

use strict;
use warnings;
use Carp 'croak';
use Exporter 'import';
use FFI::Platypus;
use FFI::Platypus::Buffer;
use FFI::Platypus::Memory;
use Time::FFI::tm;

our $VERSION = '2.003';

our @EXPORT_OK = qw(asctime ctime gmtime localtime mktime strftime strptime timegm timelocal);
our %EXPORT_TAGS = (all => \@EXPORT_OK);

my $ffi = FFI::Platypus->new(api => 1, lib => [undef], ignore_not_found => 1);
$ffi->type('record(Time::FFI::tm)' => 'tm');
my $char_size = $ffi->sizeof('char');

if (defined $ffi->find_symbol('asctime_r')) {
  $ffi->attach([asctime_r => 'asctime'] => ['tm*', 'opaque'] => 'string' => sub {
    my ($xsub, $tm) = @_;
    my $rc = $xsub->($tm, my $buf = calloc(26, $char_size));
    free $buf;
    croak "asctime: $!" unless defined $rc;
    return $rc;
  });
} else {
  $ffi->attach(asctime => ['tm*'] => 'string' => sub {
    my ($xsub, $tm) = @_;
    my $rc = $xsub->($tm);
    croak "asctime: $!" unless defined $rc;
    return $rc;
  });
}

if (defined $ffi->find_symbol('ctime_r')) {
  $ffi->attach([ctime_r => 'ctime'] => ['time_t*', 'opaque'] => 'string' => sub {
    my ($xsub, $time) = @_;
    $time = time unless defined $time;
    my $rc = $xsub->(\$time, my $buf = calloc(26, $char_size));
    free $buf;
    croak "ctime: $!" unless defined $rc;
    return $rc;
  });
} else {
  $ffi->attach(ctime => ['time_t*'] => 'string' => sub {
    my ($xsub, $time) = @_;
    $time = time unless defined $time;
    my $rc = $xsub->(\$time);
    croak "ctime: $!" unless defined $rc;
    return $rc;
  });
}

if (defined $ffi->find_symbol('gmtime_r')) {
  $ffi->attach([gmtime_r => 'gmtime'] => ['time_t*', 'tm*'] => 'opaque' => sub {
    my ($xsub, $time) = @_;
    $time = time unless defined $time;
    my $rc = $xsub->(\$time, my $tm = Time::FFI::tm->new);
    croak "gmtime: $!" unless defined $rc;
    return $tm;
  });
} else {
  $ffi->attach(gmtime => ['time_t*'] => 'tm*' => sub {
    my ($xsub, $time) = @_;
    $time = time unless defined $time;
    my $rc = $xsub->(\$time);
    croak "gmtime: $!" unless defined $rc;
    return $rc;
  });
}

if (defined $ffi->find_symbol('localtime_r')) {
  $ffi->attach([localtime_r => 'localtime'] => ['time_t*', 'tm*'] => 'opaque' => sub {
    my ($xsub, $time) = @_;
    $time = time unless defined $time;
    my $rc = $xsub->(\$time, my $tm = Time::FFI::tm->new);
    croak "localtime: $!" unless defined $rc;
    return $tm;
  });
} else {
  $ffi->attach(localtime => ['time_t*'] => 'tm*' => sub {
    my ($xsub, $time) = @_;
    $time = time unless defined $time;
    my $rc = $xsub->(\$time);
    croak "localtime: $!" unless defined $rc;
    return $rc;
  });
}

$ffi->attach(mktime => ['tm*'] => 'time_t' => sub {
  my ($xsub, $tm) = @_;
  my $rc = $xsub->($tm);
  croak "mktime: $!" if $rc == -1;
  return $rc;
});

$ffi->attach(strftime => ['opaque', 'size_t', 'string', 'tm*'] => 'size_t' => sub {
  my ($xsub, $format, $tm) = @_;
  my $max_size = length($format) * 20;
  my $buf_size = 200;
  my $rc = 0;
  my $buf;
  until ($rc != 0) {
    $rc = $xsub->($buf = realloc($buf, $buf_size * $char_size), $buf_size, $format, $tm);
    last if $buf_size > $max_size;
  } continue {
    $buf_size *= 2;
  }
  my $str = buffer_to_scalar $buf, $rc * $char_size;
  free $buf;
  return $str;
});

$ffi->attach(strptime => ['string', 'string', 'tm*'] => 'string' => sub {
  my ($xsub, $str, $format, $tm, $remaining) = @_;
  $tm = Time::FFI::tm->new unless defined $tm;
  my $rc = $xsub->($str, $format, $tm);
  croak "strptime: Failed to match input to format string" unless defined $rc;
  $tm->isdst(-1);
  $$remaining = $rc if defined $remaining;
  return $tm;
});

$ffi->attach(timegm => ['tm*'] => 'time_t' => sub {
  my ($xsub, $tm) = @_;
  my $rc = $xsub->($tm);
  croak "timegm: $!" if $rc == -1;
  return $rc;
});

$ffi->attach(timelocal => ['tm*'] => 'time_t' => sub {
  my ($xsub, $tm) = @_;
  my $rc = $xsub->($tm);
  croak "timelocal: $!" if $rc == -1;
  return $rc;
});

1;

=head1 NAME

Time::FFI - libffi interface to POSIX date and time functions

=head1 SYNOPSIS

  use Time::FFI qw(localtime mktime strptime strftime);

  my $tm = strptime '1995-01-02 13:15:39', '%Y-%m-%d %H:%M:%S';
  my $epoch = mktime $tm;
  print "$epoch: ", strftime('%I:%M:%S %p on %B %e, %Y', $tm);

  my $tm = localtime time;
  my $datetime = $tm->to_object_as_local('DateTime');

  my $tm = gmtime time;
  my $moment = $tm->to_object_as_utc('Time::Moment');

  use Time::FFI::tm;
  my $tm = Time::FFI::tm->from_object(DateTime->now);
  my $epoch = $tm->epoch_as_local;
  my $piece = $tm->to_object_as_local('Time::Piece');

=head1 DESCRIPTION

B<Time::FFI> provides a L<libffi|FFI::Platypus> interface to POSIX date and
time functions found in F<time.h>.

The L</gmtime> and L</localtime> functions behave very differently from the
core functions of the same name, as well as those exported by L<Time::Piece>,
so you may wish to call them as e.g. C<Time::FFI::gmtime> rather than importing
them.

All functions will throw an exception in the event of an error. For functions
other than L</strftime> and L</strptime>, this exception will contain the
syscall error message, and L<perlvar/$!> will also have been set by the
syscall, so you could check it after trapping the exception for finer exception
handling.

=head1 FUNCTIONS

All functions are exported individually, or with the C<:all> export tag.

=head2 asctime

  my $str = asctime $tm;

Returns a string in the format C<Wed Jun 30 21:49:08 1993\n> representing the
passed L<Time::FFI::tm> record. The thread-safe L<asctime_r(3)> function is
used if available.

=head2 ctime

  my $str = ctime $epoch;
  my $str = ctime;

Returns a string in the format C<Wed Jun 30 21:49:08 1993\n> representing the
passed epoch timestamp (defaulting to the current time) in the local time zone.
This is equivalent to L<POSIX/ctime> but uses the thread-safe L<ctime_r(3)>
function if available.

=head2 gmtime

  my $tm = gmtime $epoch;
  my $tm = gmtime;

Returns a L<Time::FFI::tm> record representing the passed epoch timestamp
(defaulting to the current time) in UTC. The thread-safe L<gmtime_r(3)>
function is used if available.

=head2 localtime

  my $tm = localtime $epoch;
  my $tm = localtime;

Returns a L<Time::FFI::tm> record representing the passed epoch timestamp
(defaulting to the current time) in the local time zone. The thread-safe
L<localtime_r(3)> function is used if available.

=head2 mktime

  my $epoch = mktime $tm;

Returns the epoch timestamp representing the passed L<Time::FFI::tm> record
interpreted in the local time zone. The time is interpreted from the C<sec>,
C<min>, C<hour>, C<mday>, C<mon>, C<year>, and C<isdst> members of the record,
ignoring the rest. DST status will be automatically determined if C<isdst> is a
negative value. The record will also be updated to normalize any out-of-range
values and populate the C<isdst>, C<wday>, and C<yday> values, as well as
C<gmtoff> and C<zone> if supported.

=head2 strftime

  my $str = strftime $format, $tm;

Returns a string formatted according to the passed format string, representing
the passed L<Time::FFI::tm> record. Consult your system's L<strftime(3)> manual
for available format descriptors.

=head2 strptime

  my $tm = strptime $str, $format;
     $tm = strptime $str, $format, $tm;
  my $tm = strptime $str, $format, undef, \my $remaining;
     $tm = strptime $str, $format, $tm, \my $remaining;

Returns a L<Time::FFI::tm> record representing the passed string, parsed
according to the passed format. Consult your system's L<strptime(3)> manual for
available format descriptors. The C<isdst> value will be set to -1; all other
unspecified values will default to 0. Note that the default C<mday> value of 0
is outside of the standard range [1,31] and may cause an error or be
interpreted as the last day of the previous month.

A L<Time::FFI::tm> record may be passed as the third argument, in which case it
will be modified in place to (on most systems) update only the date/time
elements which were parsed from the string. Additionally, an optional scalar
reference may be passed as the fourth argument, in which case it will be set to
the remaining unprocessed characters of the input string if any.

This function is usually not available on Windows.

=head2 timegm

  my $epoch = timegm $tm;

I<Since version 1.002>

Like L</mktime>, but interprets the passed L<Time::FFI::tm> record as UTC. This
function is not always available.

=head2 timelocal

  my $epoch = timelocal $tm;

I<Since version 1.002>

The same as L</mktime>, but not always available.

=head1 BUGS

Report any issues on the public bugtracker.

=head1 AUTHOR

Dan Book <dbook@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is Copyright (c) 2019 by Dan Book.

This is free software, licensed under:

  The Artistic License 2.0 (GPL Compatible)

=head1 SEE ALSO

L<Time::Piece>, L<Time::Moment>, L<DateTime>, L<POSIX>, L<POSIX::strptime>
